### DISCUSSION TITLE
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 10: character maps to <undefined>

### DISCUSSION THREAD

**Discussion by chongjing001** [sqlalchemy/sqlalchemy#8618] [Usage Questions] [answered]
### Describe the bug

 Traceback (most recent call last):
  File "./utils/result_push_tool.py", line 187, in deal_callback_data
    task_obj = task_media_obj.task
  File "/venv/lib/python3.7/site-packages/sqlalchemy/orm/attributes.py", line 276, in __get__
    return self.impl.get(instance_state(instance), dict_)
  File "/venv/lib/python3.7/site-packages/sqlalchemy/orm/attributes.py", line 682, in get
    value = self.callable_(state, passive)
  File "/venv/lib/python3.7/site-packages/sqlalchemy/orm/strategies.py", line 722, in _load_for_state
    session, state, primary_key_identity, passive
  File "<string>", line 1, in <lambda>
  File "/venv/lib/python3.7/site-packages/sqlalchemy/orm/strategies.py", line 812, in _emit_lazyload
    session.query(self.mapper), primary_key_identity
  File "/venv/lib/python3.7/site-packages/sqlalchemy/ext/baked.py", line 602, in _load_on_pk_identity
    result = list(bq.for_session(self.session).params(**params))
  File "/venv/lib/python3.7/site-packages/sqlalchemy/ext/baked.py", line 434, in __iter__
    return q._execute_and_instances(context)
  File "/venv/lib/python3.7/site-packages/sqlalchemy/orm/query.py", line 3341, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/venv/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 287, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1107, in _execute_clauseelement
    distilled_params,
  File "/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1248, in _execute_context
    e, statement, parameters, cursor, context
  File "/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1468, in _handle_dbapi_exception
    util.reraise(*exc_info)
  File "/venv/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 129, in reraise
    raise value
  File "/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1244, in _execute_context
    cursor, statement, parameters, context
  File "/venv/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
  File "/venv/lib/python3.7/site-packages/MySQLdb/cursors.py", line 206, in execute
    res = self._query(query)
  File "/venv/lib/python3.7/site-packages/MySQLdb/cursors.py", line 321, in _query
    self._post_get_result()
  File "/venv/lib/python3.7/site-packages/MySQLdb/cursors.py", line 355, in _post_get_result
    self._rows = self._fetch_row(0)
  File "/venv/lib/python3.7/site-packages/MySQLdb/cursors.py", line 328, in _fetch_row
    return self._result.fetch_row(size, self._fetch_type)
  File "/usr/local/python37/lib/python3.7/encodings/cp1252.py", line 15, in decode
    return codecs.charmap_decode(input,errors,decoding_table)
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 10: character maps to <undefined>

### To Reproduce
models.py
```python

import json
from datetime import datetime

from flask_sqlalchemy import SQLAlchemy

db = SQLAlchemy()

class BaseModel(object):
    # 定义基础的模型
    create_time = db.Column(db.DATETIME, default=datetime.now)
    update_time = db.Column(db.DATETIME, default=datetime.now, onupdate=datetime.now)

    def save(self):
        db.session.add(self)
        db.session.commit()

    def delete(self):
        db.session.delete(self)
        db.session.commit()

    def update(self):
        db.session.commit()

    # def __del__(self):
    #     db.session.refresh(self)


class Task(BaseModel, db.Model):
    __tablename__ = 'tb_task'
    id = db.Column(db.INTEGER, primary_key=True, autoincrement=True)
    task_guid = db.Column(db.String(64), unique=True)
    task_type = db.Column(db.String(16))  # 任务类型：create_song:智能上唱词，take_song:智能拍唱词，voice:智能配音,trace:智能跟踪
    name = db.Column(db.String(128))
    media_type = db.Column(db.String(16))
    pgm_guid = db.Column(db.String(64))
    doc_id = db.Column(db.String(64))
    creator_name = db.Column(db.String(64)) # 创建者姓名
    creator_code = db.Column(db.String(64)) # 创建者编号Guid
    frame_rate = db.Column(db.Float)
    track_in = db.Column(db.Integer)
    track_out = db.Column(db.Integer)
    site_id = db.Column(db.String(16))
    params = db.Column(db.Text, nullable=True)
    callback_type = db.Column(db.String(32))
    callback_address = db.Column(db.String(128))
    task_desc = db.Column(db.String(256))
    request_id = db.Column(db.String(64))
    ai_callback_task_url = db.Column(db.String(128), nullable=True)
    custom_tree = db.Column(db.String(512))
    multitype = db.Column(db.Integer, default=0)  # 1:多任务 0:单任务
    content_type = db.Column(db.Integer, default=1)
    finish_time = db.Column(db.DATETIME, default=None) # 任务完成时间
    task_status = db.Column(db.Integer, default=3) # 任务状态 1:成功 2:失败 3:处理中
    fail_msg = db.Column(db.String(256)) # 任务失败信息
    asr_clip = db.Column(db.Integer, default=0)  # 唱词srt生成入库
    extend_data = db.Column(db.Text)  # 扩展数据

    medias = db.relationship('TaskMedias', backref='task')
    voices = db.relationship('VoiceFiles', backref='task')

    def __init__(self, **kwargs):
        for key in kwargs:
            if key in ['medias', 'voice']:
                continue
            if hasattr(self, key):
                value = kwargs[key]
                if isinstance(value, list):
                    value = json.dumps(value)
                if isinstance(value, dict):
                    value = json.dumps(value)

                self.__setattr__(key, value)


class TaskMedias(BaseModel, db.Model):
    __tablename__ = 'tb_task_medias'
    id = db.Column(db.INTEGER, primary_key=True, autoincrement=True)
    clip_guid = db.Column(db.String(64))
    cliplen = db.Column(db.Integer)
    frame_rate = db.Column(db.Float)
    trackin = db.Column(db.Integer)
    trackout = db.Column(db.Integer)

    task_id = db.Column(db.Integer, db.ForeignKey('tb_task.id'))
    files = db.relationship('TaskMediaFiles', backref='task_media')


class TaskMediaFiles(BaseModel, db.Model):
    __tablename__ = 'tb_task_media_files'
    id = db.Column(db.INTEGER, primary_key=True, autoincrement=True)
    file_id = db.Column(db.String(64), unique=True)
    clipin = db.Column(db.Integer)
    clipout = db.Column(db.Integer)
    filein = db.Column(db.Integer)
    fileout = db.Column(db.Integer)
    filename = db.Column(db.String(1024))
    media_id = db.Column(db.Integer, db.ForeignKey('tb_task_medias.id'))

```

```python

import contextlib
import sqlalchemy
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.orm.attributes import InstrumentedAttribute
from mysql import connector
from api.models import db
from config import *

if DB_MODE == "1":
    engine = create_engine(f'sqlite:///{SQLITE_DB}')
else:
    engine = create_engine(f"mysql://{USER}:{PASSWORD}@{HOST}/{DATABASE}",
                           encoding='utf-8',
                           max_overflow=2,  # 超过连接池大小外最多创建的连接
                           pool_size=10,  # 连接池大小
                           pool_timeout=30,  # 池中没有线程最多等待的时间，否则报错
                           pool_recycle=POOL_RECYCLE  # 
                           )


Session = sessionmaker(bind=engine)

@contextlib.contextmanager
def get_session():
    s = Session()
    try:
        yield s
        s.commit()
    except Exception as e:
        s.rollback()
        raise e
    finally:
        s.close()

```

*in Thread*
```python

with get_session() as session:
        # ...
        file_id = results[0]['fileId']
        file_obj = session.query(TaskMediaFiles).filter_by(file_id=file_id).first()
        task_media_obj = file_obj.task_media
        task_obj = task_media_obj.task
        # ...
```

### Error

```
# Copy complete stack trace and error message here, including SQL log output if applicable.
``` 

### Versions

- OS: docker-Linux centos7
- Python: 3.7.4
- SQLAlchemy:1.3.0
- Database: mysql 5.7
- DBAPI (eg: psycopg, cx_oracle, mysqlclient):

**Comment by CaselIT** [comment]
Hi,

There is not enough information to help. What's the table sql definition and sqlalchemy type for the column?

**Comment by chongjing001** [comment]
> Hi,
> 
> There is not enough information to help. What's the table sql definition and sqlalchemy type for the column?

sorry，

**Comment by CaselIT** [comment]
Thanks for updating the example.

I'm not overly familiar with mysql (pinging @zzzeek since he is more familiar), but it seems the error is trying to store unicode in a `String` type (that's equivalent to varchar in SQL).
Try using the `Unicode` sqlalchemy type that will use nvarchar in SQL.

**Comment by zzzeek** [comment, marked as answer]
you have some cp1252 encoding in this database, as has been detected by:

     File "/usr/local/python37/lib/python3.7/encodings/cp1252.py", line 15, in decode

so you likely want to set up that [charset](https://docs.sqlalchemy.org/en/14/dialects/mysql.html#charset-selection) appropriately:

    f"mysql://{USER}:{PASSWORD}@{HOST}/{DATABASE}?charset=cp1252"

Also [create_engine.encoding](https://docs.sqlalchemy.org/en/14/core/engines.html?highlight=create_engine+encoding#sqlalchemy.create_engine.params.encoding) is not used in python 3, remove this.

**Reply by chongjing001** [reply]
Thanks
